plugin/trino-spooling-filesystem/src/main/java/io/trino/spooling/filesystem/FileSystemSpooledSegmentHandle.java

```diff
@@ -13,42 +13,98 @@
  */
 package io.trino.spooling.filesystem;
 
+import com.google.common.base.Splitter;
 import io.airlift.slice.Slice;
-import io.airlift.units.Duration;
+import io.azam.ulidj.ULID;
+import io.trino.spi.QueryId;
 import io.trino.spi.protocol.SpooledSegmentHandle;
-import io.trino.spi.protocol.SpoolingContext;
 
+import java.security.SecureRandom;
 import java.time.Instant;
-import java.time.temporal.ChronoUnit;
+import java.util.HexFormat;
+import java.util.List;
 import java.util.Optional;
 
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Verify.verify;
 import static java.util.Objects.requireNonNull;
-import static java.util.UUID.randomUUID;
 
-public record FileSystemSpooledSegmentHandle(String objectName, Instant validUntil, Optional<Slice> encryptionKey)
+public record FileSystemSpooledSegmentHandle(QueryId queryId, byte[] uuid, Optional<Slice> encryptionKey)
         implements SpooledSegmentHandle
 {
+    private static final String OBJECT_NAME_SEPARATOR = "::";
+
     public FileSystemSpooledSegmentHandle
     {
-        requireNonNull(objectName, "objectName is null");
-        validUntil = requireNonNull(validUntil, "validUntil is null").truncatedTo(ChronoUnit.MILLIS);
+        requireNonNull(queryId, "queryId is null");
+        requireNonNull(encryptionKey, "encryptionKey is null");
+        verify(uuid.length == 16, "uuid must be 128 bits");
     }
 
-    public static FileSystemSpooledSegmentHandle random(SpoolingContext context, Duration ttl)
+    public static FileSystemSpooledSegmentHandle random(SecureRandom random, QueryId queryId, Instant expireAt)
```

Comment:
Why `SecureRandom` - it can be extremly slow - I think you do not care about "real" randomness here.

Reply:
Yeah, you are right, I was overeager here